classes/mail/Mailer.inc.php

```diff
@@ -0,0 +1,196 @@
+
+        if (!PKPLocale::isLocaleValid($locale))
+            throw new InvalidArgumentException($locale . ' isn\'t recognized as a valid locale');
+
+        return array_map(function ($varValue) use ($locale) {
+
+            // Array values contain localized template data
+            if (is_array($varValue)) {
+                if (array_key_exists($locale, $varValue)) {
+                    return $varValue[$locale];
+                } else {
+                    throw new InvalidArgumentException('Template variables doesn\'t have localization for the ' . $locale . ' locale' );
+                }
+            }
+            return $varValue;
+        }, $data);
+    }
+
+    /**
+     * @param int $key number of char in a string starting from 0
+     * @param string $char
+     * @param array $template array of chars of a template
+     * @return bool
+     * @brief defines an opening tag of a variable
+     */
+    protected function openTags(int $key, string $char, array $template) : bool
+    {
+        if ($char !== self::OPEN_TAG)
+            return false;
+        if ($template[$key+1] === self::DOLLAR_SIGN_TAG)
+            return true;
+        return false;
+    }
+
+    /**
+     * @param string $compiledString the port string that has been compiled already
+     * @param string $potentialVariable candidate for a variable to seek in data
+     * @param array $data ['variable' => value]
+     * @return string string with a compiled variable
+     * @throws Exception if data doesn't include the variable
+     */
+    protected function recordVariable(string $compiledString, string $potentialVariable, array $data) : string
+    {
+        $potentialVariable = trim(ltrim($potentialVariable, self::DOLLAR_SIGN_TAG)); // makes {$ variable } and similar also allowed
+        if (array_key_exists($potentialVariable, $data)) {
+            if (filter_var($potentialVariable, FILTER_VALIDATE_URL)) {
+                return $this->handleUrl($compiledString, $potentialVariable, $data);
+            } else {
+                return $compiledString . $data[$potentialVariable];
+            }
+        }
+
+        throw new Exception('Variable with name ' . $potentialVariable . ' is not assigned to this email template');
+    }
+
+    /**
+     * @param string $compiledString the port string that has been compiled already
+     * @param string $potentialVariable URL
+     * @param array $data ['variable' => value]
+     * @return string
+     * @brief manage URLs as HTML links
+     */
+    protected function handleUrl(string $compiledString, string $potentialVariable, array $data) : string
+    {
+        $latestSymbols = substr($compiledString, -2);
+        $value = $data[$potentialVariable];
+        if ($latestSymbols === '=\'' || $latestSymbols === '=\"') {
+            return $compiledString . $value;
+        }
+
+        return $compiledString . '<a href="' . $value . '" class="' . $potentialVariable . '-style-class">' . $value . '</a>';
```

Comment:
Can you say a little bit about the use case for the `class` attribute?

Reply:
This was used originally. I wasn't able to find any but preferred to keep.

Reply:
Removed